src/morphospaces/skeleton/prediction_to_skeleton_graph.py

```diff
@@ -0,0 +1,303 @@
+    ):
+        for j in node_to_merge:
+            nx.contracted_nodes(
+                graph_merged, point_index, j + 1, copy=False, self_loops=False
+            )
+
+    for point_index, i in enumerate(point_indices_joined):
+        nx.set_node_attributes(
+            graph_merged,
+            {i: point_coordinates_joined[point_index]},
+            name="pos",
+        )
+
+    return graph_merged
+
+
+def reduce_graph(
+    graph: nx.Graph,
+    branch_point_coordinates: np.array,
+    end_point_coordinates: np.array,
+) -> nx.Graph:
+    """
+    Reduce the graph to branch- and end points. Pixels lying inbetween
+    are stored as edge attribute in an ndarray(n_pixels,3).
+
+    Parameters
+    ----------
+    graph : nx.Graph
+        pixel graph to be reduced
+    branch_point_coordinates : ndarray(n,3)
+         Array containing the point coordinates of n branch points
+    end_point_coordinates : ndarray(n,3)
+         Array containing the point coordinates of n end points
+
+
+    Returns
+    -------
+    nx.Graph
+        Reduces input graph
+    """
+
+    branch_points = []
+    end_points = []
+
+    branch_point_coordinates_s = []
+    end_point_coordinates_s = []
+
+    for node, positions in graph.nodes(data="pos"):
+        for bp in branch_point_coordinates:
+            if np.allclose(positions.astype(int), bp.astype(int)):
+                branch_points.append(node)
+                branch_point_coordinates_s.append(positions)
+        for ep in end_point_coordinates:
+            if np.allclose(positions.astype(int), ep.astype(int)):
+                end_points.append(node)
+                end_point_coordinates_s.append(positions)
+
+    point_indices_joined = branch_points + end_points
+
+    # find shortest path in image from branchpoints to endpoints
+    sps = []
+    for bp in branch_points:
+        for point in point_indices_joined:
+            if point == bp:
+                continue
+            else:
+                sps.append(nx.shortest_path(graph, bp, point))
+
+    # remove all connections including all points.
+    # Very slow, maybe do in one step?
+    for num_nodes_in_path in sps:
+        if len(np.intersect1d(num_nodes_in_path, point_indices_joined)) > 2:
+            sps.remove(num_nodes_in_path)
+
+    nodes_to_store = {}
+    node_coordinates = nx.get_node_attributes(graph, "pos")
+    for edge in sps:
+        node_to_store = [
+            node for node in edge if node not in [edge[0], edge[-1]]
+        ]
+        node_to_store = np.array([node_coordinates[n] for n in node_to_store])
+        nodes_to_store[(edge[0], edge[-1])] = node_to_store
+
+    # create the reduced graph
+    graph_reduced = nx.Graph()
+
+    for n1, n2 in nodes_to_store:
+        graph_reduced.add_edge(n1, n2)
+    nx.set_edge_attributes(graph_reduced, nodes_to_store, "points_inbetween")
+    reduced_positions = {}
+    for node in graph_reduced.nodes:
+        reduced_positions[node] = node_coordinates[node]
```

Comment:
I think the dictionary comprehension is more readable here
```suggestion
    reduced_positions = {node: node_coordinates[node] for node in graph_reduced.nodes}
```